### PR TITLE
IBM-Swift/Kitura#976 Another attempt to fix socket I/O guards

### PR DESCRIPTION
**Pull request by na-gupta**
## Description
Another attempt to fix socket I/O guards, hopefully this covers all the edge cases

## Motivation and Context
IBM-Swift/Kitura#976

## Checklist:

- [ ] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [ ] If applicable, I have updated the documentation accordingly.
- [ ] If applicable, I have added tests to cover my changes.

### CODE REVIEW

File: Sources/KituraNet/IncomingSocketHandler.swift [hunk 1]
```diff
@@ -60,10 +60,24 @@ public class IncomingSocketHandler {
...
+    /// isOpen is set to false when:
+    ///   - close() is invoked AND
+    ///   - it is safe to close the socket (there is no data waiting to be written and a socket read/write is not in progress).
+    /// This lets other threads know to not start reads/writes on this socket anymore, which could cause a crash.
     private var isOpen = true
+
+    /// writeInProgress, handleWriteInProgress and handleReadInProgress are:
+    ///   - set when the corresponding functions start their work.
+    ///   - unset when the corresponding functions finish their work.
+    /// This lets other threads know to not try to close this socket while those functions are using it, which could cause a crash.
```
Comment (shmuelk) [inline Sources/KituraNet/IncomingSocketHandler.swift]: The comment above seems to be an API comment (i.e. ///) if so, it should be repeated per variable involved.

File: Sources/KituraNet/IncomingSocketHandler.swift [hunk 2]
```diff
@@ -290,50 +334,58 @@ public class IncomingSocketHandler {
...
+            // a read/write function in between us clearing the guard and setting the flag.
+            // Make sure to set it back to open if the guard fails and we don't actually close.
+            guard !writeInProgress && !handleWriteInProgress && !handleReadInProgress
+                && writeBuffer.length == writeBufferPosition else {
+                    isOpen = true
+                    return
+            }
+
+            #if os(OSX) || os(iOS) || os(tvOS) || os(watchOS) || GCD_ASYNCH
+                readerSource.cancel()
```
Comment (shmuelk) [inline Sources/KituraNet/IncomingSocketHandler.swift]: Did you verify what happens if you call readerSource.cancel() multiple times? That is, is handleCancel() called each time readerSource.cancel() is called?

Comment (na-gupta) [inline Sources/KituraNet/IncomingSocketHandler.swift]: The handler is called just the first time like you had thought. That's why I moved the guards from the handleRead() to close()

Comment (shmuelk) [inline Sources/KituraNet/IncomingSocketHandler.swift]: The question is about **handleCancel()**, because handleCancel() closes the socket.

Comment (shmuelk) [inline Sources/KituraNet/IncomingSocketHandler.swift]: Never mind. I now understand what you meant.

Comment (na-gupta) [inline Sources/KituraNet/IncomingSocketHandler.swift]: Yes, sorry, I meant handleCancel, not handleRead. My fingers and brain don't communicate well before my morning coffee.

File: Sources/KituraNet/IncomingSocketHandler.swift (hunk 1 above)
Comment (na-gupta) [inline Sources/KituraNet/IncomingSocketHandler.swift]: Fixed so we only use API style comments for public/open members.

**Pull request merged**